test/dtls_mtu_test.c

```diff
@@ -0,0 +1,213 @@
+				      unsigned char *psk,
+				      unsigned int max_psk_len)
+{
+	snprintf(ident, max_ident_len, "psk");
+
+	if (max_psk_len > 20)
+		max_psk_len = 20;
+	memset(psk, 0x5a, max_psk_len);
+
+	return max_psk_len;
+}
+
+static unsigned int srvr_psk_callback(SSL *ssl, const char *identity,
+				      unsigned char *psk,
+				      unsigned int max_psk_len)
+{
+	if (max_psk_len > 20)
+		max_psk_len = 20;
+	memset(psk, 0x5a, max_psk_len);
+	return max_psk_len;
+}
+
+static int mtu_test(SSL_CTX *ctx, const char *cs, int no_etm)
+{
+	SSL *srvr_ssl = NULL, *clnt_ssl = NULL;
+	BIO *sc_bio = NULL, *cs_bio = NULL;
+	int ret;
+	int i, j;
+	int mtus[30];
+	unsigned char buf[530];
+	int rv = 0;
+
+	memset(buf, 0x5a, sizeof(buf));
+
+	sc_bio = BIO_new(BIO_s_mem());
+	cs_bio = BIO_new(BIO_s_mem());
+
+	/* Non-blocking */
+	BIO_set_mem_eof_return(sc_bio, -1);
+	BIO_set_mem_eof_return(cs_bio, -1);
+
+	clnt_ssl = SSL_new(ctx);
+	srvr_ssl = SSL_new(ctx);
+
+	if (no_etm)
+		SSL_set_options(srvr_ssl, SSL_OP_NO_ENCRYPT_THEN_MAC);
+
+	if (SSL_set_cipher_list(srvr_ssl, cs) != 1 ||
+	    SSL_set_cipher_list(clnt_ssl, cs) != 1) {
+		ERR_print_errors_fp(stdout);
+		goto out;
+	}
+
+
+	/* The BIOs very freed twice, once from each SSL. Compensate. */
+	BIO_up_ref(sc_bio);
+	BIO_up_ref(cs_bio);
+
+	SSL_set_bio(clnt_ssl, cs_bio, sc_bio);
+	SSL_set_bio(srvr_ssl, sc_bio, cs_bio);
+
+	while (1) {
+		/* Client first, then server. And client finishes last */
+		ret = SSL_connect(clnt_ssl);
+		if (ret == 1)
+			break;
+		else if (SSL_get_error(clnt_ssl, ret) != SSL_ERROR_WANT_READ) {
+			ERR_print_errors_fp(stdout);
+			goto out;
+		}
+
+		ret = SSL_accept(srvr_ssl);
+		if (ret == 1) {
+			if (debug)
+				printf("Server established connection\n");
+		} else if (SSL_get_error(srvr_ssl, ret) != SSL_ERROR_WANT_READ) {
+			ERR_print_errors_fp(stdout);
+			goto out;
+		}
+
+	}
+	if (debug)
+		printf("Channel established\n");
+
+	/* For record MTU values between 500 and 539, call DTLS_get_data_mtu()
+	 * to query the payload MTU which will fit. */
+	for (i = 0; i < 30; i++) {
+		SSL_set_mtu(clnt_ssl, 500 + i);
+		mtus[i] = DTLS_get_data_mtu(clnt_ssl);
+		if (debug)
+			printf("%s%s payload MTU for record mtu %d = %d\n",
+			       cs, no_etm ? "-noEtM":"", 500 + i, mtus[i]);
+		if (mtus[i] == 0) {
+			fprintf(stderr,
+				"payload MTU failed with record MTU %d for %s\n",
+				500 + i, cs);
+			goto out;
+		}
+	}
+
+	/* Now get out of the way */
+	SSL_set_mtu(clnt_ssl, 1000);
+
+	/* Now for all values in the range of payload MTUs, send
+	 * a payload of that size and see what actual record size
+	 * we end up with. */
+	j = 0;
+	for (i = mtus[0]; i <= mtus[29]; i++) {
+		int reclen;
+		if (SSL_write(clnt_ssl, buf, i) != i) {
+			ERR_print_errors_fp(stdout);
+			goto out;
+		}
+		reclen = BIO_pending(sc_bio);
+		(void)BIO_reset(sc_bio);
+		if (debug)
+			printf("record %d for payload %d\n", reclen, i);
+
+		for (j = 0; j < 30; j++) {
+			if (i <= mtus[j] && reclen > (500 + j)) {
```

Comment:
Not with CBC. For example, a record size (`SSL_set_mtu()`) of anything between (j+500=) 503-519 might *all* give a data MTU of (i=) 477. 

I'm fairly sure the two tests here *are* what we need to check for. There was plenty of caffeine involved, as well as empirical testing and even some bug fixes which I silently reworked into the `DTLS_get_data_mtu()` implementation.

They test that sending a given payload of size `i` will never result in a DTLS record larger than **any** of the record MTU values for which a max payload size <= `i` was reported, and that a given payload of size `i` will never result in a DTLS record which is *smaller* than any of the record MTU values for which a max payload size > `i` was reported.

The former for obvious reasons, the latter because it would mean that `DTLS_get_data_mtu()` is reporting a value which is too low -- it *does* fit in the requested record MTU, but *more* could fit too.